### PR TITLE
feature: support DbString in DapperAOT

### LINKED ISSUE #89: DbString - analyzer and runtime
`DbString` is a Dapper type that allows metadata to be conveyed for string types.

However:

- it is allocatey
- it isn't touched by AOT today
- what it achieves is the same as [`DbValue]`](https://github.com/DapperLib/DapperAOT/blob/main/src/Dapper.AOT/DbValueAttribute.cs)

Suggestions:

1. we need to verify that `[DbValue]` allows the size and type to be conveyed into `DbParameter`
2. we should emit a warning to use `[DbValue]` instead
3. we should probably treat `DbString` as a recognized type, and configure things correctly
4. for this last, I would propose (if any `DbString` are encountered) writing a helper method into the output (we have some prior art that copies entire files into the output; that would be fine) with a method that takes a `DbParameter` and `DbString` (`global::` qualified) and does the right things:

- if the entire `DbString` is null, just set `Value` to `DbNull`
- set the size if appropriate
- set the type if appropriate
- set the value using `AsValue`, or a `(object)value.Value ?? DbNull` if easier

The reason we can't put this into the library is that we can't ref Dapper directly because of the Dapper/Dapper.StrongName duality - hence we need to implement it in codegen

### PR DESCRIPTION
**Pull request by DeagleGross**
`DbString` is a known type in [Dapper](https://github.com/DapperLib/Dapper/blob/main/Dapper/DbString.cs) and currently there is no dedicated support in DapperAOT for it.

Current PR does 2 things regarding `DbString`:
1) in analyzer mode reports DAP48 and suggests to move from usage of `DbString` to `DbValue` attribute.
2) properly configures `DbParameter` if one of parameter members is `DbString` by setting `.Value`, `.Size` and `.DbType`

Implementation notes:
---
1) I have moved files in `Dapper.AOT.Analyzers` which are copied to the output in separate folder `InGeneration`
2) I have created a single code writer for such types (`PreGeneratedCodeWriter`), and included the generation context options in `GenerateState` to have a control over which pre-generated types to include in the output. I decided to do in order to avoid writing `DbStringHelpers` in the output even in those cases, where `DbString` is not even used (basically to have more clear output)

Questions:
---
1) I am pretty sure I messed up on how to get the size of DbString in ANSI encoding. How do we do it properly?
2) What exactly do we need to do with the `DbString` in `UpdateParameters`? Do we need to call the same setup there, or just re-assigning value is fine?

Closes #89

### CODE REVIEW

File: test/Dapper.AOT.Test/Verifiers/DAP048.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+using System.Threading.Tasks;
+using Xunit;
+using static Dapper.CodeAnalysis.DapperAnalyzer;
+
+namespace Dapper.AOT.Test.Verifiers;
+
+public class DAP048 : Verifier<DapperAnalyzer>
+{
+    [Theory]
+    [InlineData("[DapperAot(false)]")]
```
Comment (DeagleGross) [inline test/Dapper.AOT.Test/Verifiers/DAP048.cs:11]: we need to emit warning ONLY in DapperAOT mode

**Comment by DeagleGross** [comment]
add Integration tests for dbstring:
- dapper.vanilla without any interception (no dapper aot)
- dapper code with dapper aot with dbString
- dbvalue (interception)

File: test/Dapper.AOT.Test/Verifiers/DAP048.cs (hunk 1 above)
Comment (DeagleGross) [reply test/Dapper.AOT.Test/Verifiers/DAP048.cs]: done

File: src/Dapper.AOT.Analyzers/Dapper.AOT.Analyzers.csproj [hunk 2]
```diff
@@ -16,9 +16,11 @@
 
     <ItemGroup>
         <Compile Remove="AotGridReader.cs" />
-        <Compile Remove="InterceptsLocationAttribute.cs" />
+        <Compile Remove="InGeneration\DapperHelpers.cs" />
```
Comment (mgravell) [inline src/Dapper.AOT.Analyzers/Dapper.AOT.Analyzers.csproj:19]: maybe wildcard the folder for compile/embeddedresource?

**Review by mgravell — APPROVED** [review]
This is looking great; there's a ton of work in here, thanks; a couple of minor optional nits, but: massive thanks

File: docs/rules/DAP048.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
+﻿# DAP048
+
+[DbString](https://github.com/DapperLib/Dapper/blob/main/Dapper/DbString.cs) is allocatey type, but achieves the same as
```
[the remark this answers is not in the document]
Comment (DeagleGross) [reply docs/rules/DAP048.md]: changed

File: src/Dapper.AOT.Analyzers/Dapper.AOT.Analyzers.csproj (hunk 2 above)
Comment (DeagleGross) [reply src/Dapper.AOT.Analyzers/Dapper.AOT.Analyzers.csproj]: sure, made a wildcard here

**Comment by DeagleGross** [comment]
Hey @mgravell,
I know you asked me to implement integration tests - I tried but I still dont have a solution yet. Filed an issue to do it in a separate PR: #122

regarding testing `DbString` functionality: I packed Dapper.AOT, installed in local NuGet feed and used in a console app:
1) I can confirm that we emit a proposition to migrate from `DbString` to `[DbValue]` (only in Dapper.AOT mode)
[module: DapperAot]
image

2) confirmed that query against local database works and returns the model I have in database (_note: we hit the breakpoint in the generated `DbStringHelpers` - look at screenshot_):
image

[Repro of DbString usage](https://gist.github.com/DeagleGross/c36cb63f1d8f64457fa29a61c6f33c90)

File: src/Dapper.AOT.Analyzers/Internal/Inspection.cs [hunk 4]
```diff
@@ -488,6 +495,23 @@ public string DbName
...
+        public DapperSpecialType DapperSpecialType 
+        { 
+            get 
+            {
+                if (CodeType is {
+                    Name: "DbString",
+                    TypeKind: TypeKind.Class,
+                    ContainingNamespace:
+                    {
+                        Name: "Dapper"
```
Comment (mgravell) [inline src/Dapper.AOT.Analyzers/Internal/Inspection.cs:507]: nit: ContainingNamespace.IsGlobalNamespace: true ?

**Review by mgravell — APPROVED** [review]

**Pull request merged**